control-plane/agents/src/bin/core/volume/service.rs

```diff
@@ -348,19 +351,52 @@ impl Service {
             .await?;
 
         let spec = snapshot.as_ref().spec();
-        let info = VolumeSnapshotInfo::new(spec.source_id(), spec.uuid().clone());
+        let info = CreateVolumeSnapshot::new(spec.source_id(), spec.uuid().clone());
         // todo: get state from registry..
         let state = VolumeSnapshotState::new(info, 0, None);
         Ok(VolumeSnapshot::new(snapshot.as_ref().into(), state))
     }
 
     /// Delete a volume snapshot.
-    async fn delete_snapshot(&self, info: VolumeSnapshotInfo) -> Result<(), SvcError> {
-        let mut volume = self.specs().volume(&info.source_id).await?;
+    #[tracing::instrument(level = "info", skip(self), err, fields(volume.uuid = ?request.source_id, snapshot.source_uuid = ?request.source_id, snapshot.uuid = %request.snap_id))]
+    async fn delete_snapshot(&self, request: DeleteVolumeSnapshot) -> Result<(), SvcError> {
+        // Fetch the snapshot spec.
+        let snapshot = self.specs().volume_snapshot_rsc(request.snap_id()).ok_or(
+            SvcError::VolSnapshotNotFound {
+                snap_id: request.snap_id().to_string(),
+                source_id: request.source_id().as_ref().map(|id| id.to_string()),
+            },
+        )?;
+        let persisted_snap_info = snapshot.lock().clone();
+
+        // Fetch the volume using the snapshot source.
+        let mut volume = match request.source_id() {
+            None => {
+                self.specs()
+                    .volume(persisted_snap_info.spec().source_id())
+                    .await
+            }
+            Some(vol_id) => {
+                if persisted_snap_info.spec().source_id() == vol_id {
+                    self.specs().volume(vol_id).await
+                } else {
+                    // Source id did not match, different snapshot.
+                    Err(SvcError::VolSnapshotNotFound {
```

Comment:
hmm seems like we should have a different error here to highlight the conflict?

Reply:
How about `InvalidArgument` with `source_id` as invalid?

Reply:
Seems good, and please make it obvious in the error message what the "expected" source id was.